### PR TITLE
Fix (CORE) transforming of external formats to scripts

### PR DESCRIPTION
**Pull request by z3dev**
This fixes issues when converting external formats (STL, DXF, etc) to JSCAD scripts.

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

### CODE REVIEW

**Comment by z3dev** [comment]
FYI, there still some changes for the other external formats required. But I would like feedback on these changes first.

File: packages/core/code-loading/transformSources.js [hunk 1]
```diff
@@ -12,24 +12,63 @@ const modulifyTransform = (options, entry) => {
...
-  const deserializeStl = require('@jscad/io').stlDeSerializer.deserialize
-  const source = deserializeStl(entry.source, entry.name, options)
-  // const ext = 'js'
-  // const name = entry.name.split('.') + ext
-  // const fullPath = entry.fullPath.split('.') + ext
-  return Object.assign({}, entry, { source })
+/*
+ * Create a new entry for a script (JSCAD) from the given entry and source
+ */
+const createJscadEntry = (entry, source) => {
```
Comment (kaosat-dev) [inline packages/core/code-loading/transformSources.js]: very nice  !

File: packages/core/code-loading/transformSources.js [hunk 2]
```diff
@@ -12,24 +12,63 @@ const modulifyTransform = (options, entry) => {
...
+  const ext = 'jscad'
+  const name = entry.name.substring(0, entry.name.lastIndexOf('.') + 1) + ext
+  const fullPath = '/' + name
+
+  return Object.assign({}, entry, { ext, name, fullPath, source })
+}
+
+const transformAmfToJscad = (options, entry) => {
```
Comment (kaosat-dev) [inline packages/core/code-loading/transformSources.js]: these are all great !
With a big catch, please read my comment in the overral review

**Comment by z3dev** [comment]
> * do we use CSG based deserialization ?

Wow! I’m not sure... so I suggest we start by allowing the user to choose.

In the website, I guess this means another ‘option’.

Maybe this also all applies to the application.

For the CLI, maybe there could be more granularity, allowing the user to choose which formats to convert into what.

**Comment by z3dev** [comment]
By the way. I like the ‘registration’ of extensions in the core, which is sourced from the IO library. But it’s hard coded to ‘csg’ now. Maybe this is where the new ‘option’ can be useful, allowing ‘jscad’ or ‘csg’

**Comment by kaosat-dev** [comment]
> > * do we use CSG based deserialization ?
> Wow! I’m not sure... so I suggest we start by allowing the user to choose.
> In the website, I guess this means another ‘option’.
> Maybe this also all applies to the application.
> For the CLI, maybe there could be more granularity, allowing the user to choose which formats to convert into what.

I will be honest Jeff, the issue is even more complex than that, a simple option is not enough because in certain cases it would not work (one file vs multiple, external vs internal editor etc). For now I think we should leave it as is, and add a note/ issue for it, because it is HARD to solve (I went through quite a few iterations trying to solve it, and none was satisfactory).
The best solution I found so far:
- if you drop or work on a **SINGLE** file, then do a text/jscad conversion and display that
- if you drop **multiple files or a FOLDER** then use csg as you are likely (as a user) to want to work on a full project where you can import your stl/amf/whatever files as CSG using something like ```shape = require('./mySTL.STL')

**Comment by kaosat-dev** [comment]
> By the way. I like the ‘registration’ of extensions in the core, which is sourced from the IO library. But it’s hard coded to ‘csg’ now. Maybe this is where the new ‘option’ can be useful, allowing ‘jscad’ or ‘csg’

sorry not sure I follow ?

**Comment by z3dev** [comment]
> > By the way. I like the ‘registration’ of extensions in the core, which is sourced from the IO library. But it’s hard coded to ‘csg’ now. Maybe this is where the new ‘option’ can be useful, allowing ‘jscad’ or ‘csg’
> 
> sorry not sure I follow ?

OK. I looks like the webrequire is calling a routine call 'registerAllExtensions' found in core/io

This seems like the correct way to go, rather then having each application build another 'registration' of formats.

I really think this should be part of the IO library though, as that library is the source of the deserializers / serializers. It would be trivial to make this the source of the external 'format' information as well.

**Comment by kaosat-dev** [comment]
> I really think this should be part of the IO library though, as that library is the source of the deserializers / serializers. It would be trivial to make this the source of the external 'format' information as well.

Sorry but no :) this USES the code from IO to handle the format, but the registration process is specific to the 'core'/ main script evaluation process.

**Comment by z3dev** [comment]
> Sorry but no :) this USES the code from IO to handle the format, but the registration process is specific to the 'core'/ main script evaluation process.

Understood. The core decides how to handle each request to process the external format. But the information about each format should come from the IO library.

```
  amf: {
    displayName: 'AMF (experimental)',
    description: 'Additive Manufacturing File Format',
    extension: 'amf',
    mimetype: 'application/amf+xml',
    convertCSG: true,
    convertCAG: false },
```

**Comment by kaosat-dev** [comment]
The format description that you mention is already in IO :) (formats.js)

**Pull request merged**